### DISCUSSION TITLE
Livestream with Tyler Leonhardt on Extension Development!

### DISCUSSION THREAD

**Discussion by olguzzar** [microsoft/vscode-discussions#707] [Announcements]
Hi all! In about 15 minutes (8 AM Pacific Time on July 20th), we'll be livestreaming from our VS Code YouTube channel with Tyler Leonhardt, an engineer on the VS Code team that many of you probably know. He'll be doing a live coding session on extension development, covering how to test an extension and how to make an extension run in the web. Check it out, and the replay, [here](https://youtube.com/live/4WblV-wbih8)!

**Comment by r3m0t** [comment]
It's a good demo of how difficult it is to test extensions! Just to wait for a quick pick to open, he called an undocumented command `getContextKeyInfo` , waited for `inQuickOpen` (which is misnamed- isn't Quick Open the `Ctrl+P` feature listing files?),  sleeped an arbitrary 100 milliseconds, then had to call more undocumented commands to interact with the quick pick.

And I know there's some VS Code features that don't work in tests at all like the modal dialog boxes.

It would be great if these tools were wrapped up in a package or a VSCode API that is only available in ExtensionMode of test or development. The stability of all extensions would benefit by being able to test against new/insider versions without mocking out, eg the entire QuickPick API or the entire `TextEditor`.

**Reply by r3m0t** [reply]
My code is setting `quickPick.value` and `quickPick.valueSelection`, is there any way to test it is showing on screen? My code updates `quickPick.items` when `quickPick.onDidChangeValue`, is there a command to execute to type into the quick pick and assert the items displayed? etc. This whole approach just breaks down really quickly IMO.

**Comment by alefragnani** [comment]
For those looking for the source code of the [Fugio](https://marketplace.visualstudio.com/items?itemName=TylerLeonhardt.fugio) extension that Tyler created for the stream, you can find it here https://github.com/TylerLeonhardt/vscode-fugio-stream. It's a modified version from the one you find in the marketplace.

Another extension I suggest as a good source about testing is [Peacock](https://marketplace.visualstudio.com/items?itemName=johnpapa.vscode-peacock) from John Papa. When I decided to play with testing my own extensions, for things other than unit tests, Peacock source was a great guide.

It's great to see more references and samples about testing.
